### ISSUE TITLE
Sorting based on linked properties

### ISSUE THREAD

**Issue by Taco55** [realm/realm-dart#1183]
### Problem

I would like to sort a list based on linked properties. Would that be possible? 

I have seen a possibly related issue for Realm Java [#672](https://github.com/realm/realm-java/issues/67), and which appeared to be resolved in PR [#3834](https://github.com/realm/realm-java/pull/3834) in 2017. However, I don't see information in the documentation on how sorting on linked properties would workout.

In my case, I would like to sort a list of`Item`s, in which each item has a one-to-many relationship with `FieldValue`. The sorting, however, should be performed on a property in a single `FieldValue`. 

In pure Dart I would do it like this: 

```
import 'package:collection/collection.dart';

part 'main.g.dart';

@RealmModel()
class _Item {
  @MapTo('_id')
  @PrimaryKey()
  late ObjectId id;
  late List<_FieldValue> fieldValues = const [];
}

@RealmModel()
class _FieldValue {
  late _FieldProperty? fieldProperty;
  late String? textValue;
}

@RealmModel()
class _FieldProperty {
  @MapTo('_id')
  @PrimaryKey()
  late ObjectId id;

  String name = "";
}

void main(List<String> arguments) {
  final realm = Realm(Configuration.local([
    FieldValue.schema,
    FieldProperty.schema,
    Item.schema,
  ]));

  final field = FieldProperty(ObjectId(), name: "text");
  final unsortedItems = [
    Item(ObjectId(), fieldValues: [FieldValue(ObjectId(), textValue: "a", fieldProperty: field)]),
    Item(ObjectId(), fieldValues: [FieldValue(ObjectId(), textValue: "d", fieldProperty: field)]),
    Item(ObjectId(), fieldValues: [FieldValue(ObjectId(), textValue: "c", fieldProperty: field)]),
    Item(ObjectId(), fieldValues: [FieldValue(ObjectId(),                 fieldProperty: field)]),
    Item(ObjectId(), fieldValues: [FieldValue(ObjectId(), textValue: "b", fieldProperty: field)])
  ];
  realm.write(() {
    realm.add(field);
    realm.addAll(unsortedItems);
  });

  var unsortedTextValues = unsortedItems
      .map((item) => (item.fieldValues
              .firstWhereOrNull((element) => element.fieldProperty == field)
              ?.textValue ??
          ""))
      .join(", ");

  print(unsortedTextValues); // a, d, c, , b

  // Sort items
  var sortedItems = unsortedItems.sorted((a, b) => (a.fieldValues
              .firstWhereOrNull((element) => element.fieldProperty == field)
              ?.textValue ??
          "")
      .compareTo((b.fieldValues.firstWhereOrNull(
              (element) => element.fieldProperty == field))?.textValue ??
          ""));

  var sortedTextValues = sortedItems
      .map((item) => (item.fieldValues
              .firstWhereOrNull((element) => element.fieldProperty == field)
              ?.textValue ??
          ""))
      .join(", ");

  print(sortedTextValues); //  , a, b, c, d

  Realm.shutdown();
}
```

Is this also possible with Realm Query Language?

### How important is this improvement for you?

Would be a major improvement

### Feature would mainly be used with

Local Database only

**Comment by nielsenko** [comment]
The semantics are subtly different, but would:
```dart
  final sortedValues = realm.query<FieldValue>(r'fieldProperty == $0 SORT(textValue ASC)', [field]);
  print(sortedValues.map((v) => v.textValue ?? ''));
```
work for you?

Outputs:
```
(, a, b, c, d)
```

**Comment by Taco55** [comment]
@nielsenko 
This surely helps. 
However, in my case some more logic is needed since I need to have `Item`s sorted instead of `FieldValue`s. 

I would need something like this:

```
  final sortedItems = realm.query<Item>(
      r'fieldValues.fieldProperty == $0 SORT(fieldValues.textValue ASC)',
      [field]);

```
But this does gives the following error:
`Exception has occurred.
RealmException (RealmException: Error code: 18 . Message: Data type mismatch`

Would it still be possible to sort in this way?

**Comment by nielsenko** [comment]
Yes, what does it mean to sort by a collection? `fieldValues` is a collection. I don't think we can currently express that in RQL.

**Comment by Taco55** [comment]
`FieldValues` is a collection indeed, but sorting is only based on one property at a time in this list (in my example it is `textValue`, but it can also be any other value such as a `date`). 
There were good reasons for choosing this data model and there weren't really any other satisfying options. Quite complex filtering is possible with RQL, but the sorting possibilities appear to be more limited. I think I should have looked into that better in advance. 
I realize that the current functionality might be sufficient in most cases, but for this use case it's a pity. Since this app relies on the ability to filter and sort based on this data model, the only solution I see right now is to convert all data to normal lists and use Dart sorting. Not what I really want of course...
I really hope more sorting options will come available in the near future.

**Comment by nielsenko** [comment]
Something like:
```dart
 realm.query<Item>(r'$v IN fieldValues ANY $v.fieldProperty == $0 SORT($v.textValue ASC)')
```
(just speculating)

I have no idea how difficult this would be, but I have a hunch it won't be trivial. 

I think @astigsen might be interested in this feature request, so I'll tag him here.

**Comment by Taco55** [comment]
This would be a great outcome for me at least if this could become possible. I'm looking forward to a possible analysis on this. 
For now it is a dilemma how to deal with the current scheme choice.

**Comment by nirinchev** [comment]
Do you need to get change notifications for the query or is it a one-off query? Depending on your data size, you could also perform the filtering/sorting in-memory.

**Comment by Taco55** [comment]
@nirinchev I do make use of change notifications, but at this moment I did not see other options than converting the data to a normal list and to sort the list in Dart. In this case remote changes will trigger a complete rebuild. Is this also what you mean?

**Comment by nirinchev** [comment]
Yes, that was the unspoken concern in my last message - converting to a regular list and using dart sorting should work, but you're going to lose out on the detailed change information.

**Comment by Taco55** [comment]
Indeed, I would rather just use RQL. Could it be that RQL will offer more possibilities for sorting in the near future? Or is that not to be expected?

**Comment by nirinchev** [comment]
We don't have anything of the sort planned for the near/mid-term future, but it's a worthwhile improvement that we hope to get to at some point.
